src/hierarchy_builder.rs

```diff
@@ -0,0 +1,363 @@
+                    candidate.zone_type,
+                    smallest.and_then(|s: &Zone| s.zone_type),
+                ) && candidate.contains(z)
+                {
+                    Some(candidate)
+                } else {
+                    smallest
+                }
+            });
+
+        z.set_parent(parent.map(|z| z.id.clone()));
+    }
+}
+
+// This struct is necessary to wrap the `zones` slice
+// and keep a mutable reference to a zone (and set
+// its parent) while still be able to borrow another
+// reference to another zone.
+struct MutableSlice<'a> {
+    pub right: &'a [Zone],
+    pub left: &'a [Zone],
+    pub idx: usize,
+}
+
+impl<'a> MutableSlice<'a> {
+    pub fn init(zones: &'a mut [Zone], index: usize) -> (Self, &'a mut Zone) {
+        let (left, temp) = zones.split_at_mut(index);
+        let (z, right) = temp.split_at_mut(1);
+        let s = Self {
+            right: right,
+            left: left,
+            idx: index,
+        };
+        (s, &mut z[0])
+    }
+
+    pub fn get(&self, zindex: &ZoneIndex) -> &Zone {
+        let idx = zindex.index;
+        if idx < self.idx {
+            return &self.left[idx];
+        } else if idx == self.idx {
+            panic!("Cannot retrieve middle index");
+        } else {
+            return &self.right[idx - self.idx - 1];
+        }
+    }
+}
+
+// the goal is that f in [down(f as f32) as f64, up(f as f32) as f64]
+fn down(f: f32) -> f32 {
+    f - (f * ::std::f32::EPSILON).abs()
+}
+fn up(f: f32) -> f32 {
+    f + (f * ::std::f32::EPSILON).abs()
+}
+
+#[cfg(test)]
+mod test {
+    use geo::{LineString, MultiPolygon, Point, Polygon};
+    use zone::{Zone, ZoneIndex, ZoneType};
+    use hierarchy_builder::build_hierarchy;
+    use osmpbfreader::Tags;
+
+    fn zone_factory(idx: usize, ls: LineString<f64>, zone_type: Option<ZoneType>) -> Zone {
+        let p = Polygon::new(ls, vec![]);
+        let mp = MultiPolygon(vec![p.clone()]);
+
+        Zone {
+            id: ZoneIndex { index: idx },
+            osm_id: "".into(),
+            admin_level: None,
+            zone_type: zone_type,
+            name: "".into(),
+            center: None,
+            boundary: Some(mp),
+            parent: None,
+            tags: Tags::new(),
+            wikidata: None,
+            zip_codes: vec![],
+        }
+    }
+
+    #[test]
+    fn test_is_smaller_simple() {
+        let smallest = Some(ZoneType::City);
+        let candidate = Some(ZoneType::Country);
+        assert_eq!(super::is_smaller(smallest, candidate), true);
+    }
+    #[test]
+    fn test_is_smaller_first_none() {
+        let smallest = None;
+        let candidate = Some(ZoneType::Country);
+        assert_eq!(super::is_smaller(smallest, candidate), false);
+    }
+    #[test]
+    fn test_is_smaller_second_none() {
+        let smallest = Some(ZoneType::City);
+        let candidate = None;
+        assert_eq!(super::is_smaller(smallest, candidate), true);
```

Comment:
Not sure to understand why `City` should be smaller than None. A comment would be useful.